### PR TITLE
sys/atomic_utils: Functions for atomic access

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

This PR provides a set of functions to perform atomic accesses. These functions closely resemble the C11 atomic library functions, but are implemented more efficiently:

- Lock-free atomic read-modify-write operations are usually slower than just disabling IRQs, even on big fat STM32F7 CPUs. Still, the compiler will assume lock-free implementations to always be the fastest option. This assumption does not hold on bare metal embedded targets.
- If no lock-free implementation is used, the compiler will emit a library call. Just disabling IRQs inline is more efficient (in terms of CPU cycles and ROM)
- In special cases (e.g. only one thread is writing to a variable), only the write part of the read-modify-write operation needs to be atomically. This can be implemented cheaper than full atomic operations and is provided here.
    - This is useful to fix https://github.com/RIOT-OS/RIOT/issues/9882
- Under certain conditions (e.g. when IRQs are disabled), mixing atomic accesses with non-atomic accesses is possible and faster. This library allows for this.
- Atomic operations for setting and clearing bits are provided
    - Unlike `atomic_fetch_or(&mask, 1 << bit)` or `atomic_fetch_and(&mask, ~(1 << bit))` this can be implemented lock-free and efficiently with bit-banding on Cortex M3 / M4 MCUs

### Testing procedure

Note: This currently only works for CortexM MCUs.

#### Performance Check
- Run the provided benchmark in `tests/bench_atomic_util`
- It contains a `README.md`. If the given documentation is not sufficient, I really should fix it there rather than here

##### Arithmetic Correctness Check
- Run the added unit tests: `BOARD=<SOME_CORTEX_M_BOARD> make -C tests/unittests/ tests-atomic_utils flash test`
    - This will be most interesting for Cortex M3, M4, and M4F boards

#### Testing "Atomic-ness"

- Run the test in `tests/sys_atomic_utils`
    - You can use `make test`, but you might want to manually test as well
    - The `README.md` and the in-application documentation should be sufficient. If not, I really should fix it there rather than here.

### Issues/PRs references

Useful for https://github.com/RIOT-OS/RIOT/issues/9882

### CODE REVIEW

**Comment by kaspar030** [comment]
Nice one!

Should we already add atomic single bit functions? `atomic_set_bit_u32(uint32_t *var, bit)` etc.
We can then easily hook them up to use bit banding if a platform supports it.

**Comment by maribu** [comment]
> We can then easily hook them up to use bit banding if a platform supports it.

Great idea. Do you have any good documentation for this? E.g. the few examples for using this I found that didn't use the bit-banding periph region did never allocate memory in the SRAM bit-banding region, but just created a wild pointer pointing somewhere in there. That might not be the best idea...

I think there should be some `__attribute__((region("bit-banding")))` magic or similar to make sure a variable actually gets placed in there. But I didn't find anything the like. If such exists, the API should also provide an attribute for bitmasks, which by default is an empty preprocessor macro.

The alternative would be to check within every `atomic_set_bit_u<WIDTH>()` function if the address of the target mask is within the SRAM bit banding region and else fall back to `atomic_fetch_or()`. But I much rather add a precondition to the function that bitmask are allocated with the magic keyword, so that users get predictable and consistent performance.

**Comment by maribu** [comment]
I think it is `__attribute__((section(''.srambb'')))` :-) Let me check if this works without touching the linker scripts

**Comment by maribu** [comment]
@kaspar030: I implemented the atomic set_bit / clear_bit functions as suggested using bit banding. Thanks for pointing me to this.

I added some unit tests for arithmetic correctness of the atomic functions; which directly has proven to be worth the effort.

**Comment by maribu** [comment]
I added one more test, which currently is only applicable to the `atomic_fetch_<OP>_u<BITS>()` family of functions. It also contains a broken implementation `volatile_fetch_<OP>_u<BITS>()` for reference, which can also be tested. The test seems to perform quite well to detect issues with the `volatile_fetch_<OP>_u<BITS>()` functions and does not detect any issues for the `atomic_*()` family of functions :-)

I still need to extend the test to also allow checking the `atomic_set_bit_u<BITS>()` and `atomic_clear_bit_u<BITS>()` functions. Also, some documentation will be needed for the test to be actually useful.

**Comment by maribu** [comment]
This is now ready for review.

**Comment by maribu** [comment]
Now that it compiles on Murdock, let's also address Travis comments, and then finally squash.

**Comment by maribu** [comment]
Finally, travis is happy. Let's see if Murdock is also still happy.

**Comment by maribu** [comment]
Hooray. All bots happy. Now the humans can give their feedback :-)

File: cpu/arm7_common/include/atomic_utils_arch.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
+/*
```
Comment (kaspar030) [inline cpu/arm7_common/include/atomic_utils_arch.h:1]: These atomic_utils_arch look awfully similar. Quick look shows AVR8 only has a subset, and some have extra bitband.
Would this be easy to dedup a bit more?

File: tests/sys_atomic_utils/volatile_utils.h [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+/**
+ * @ingroup     tests
+ * @{
+ *
+ * @file
+ * @brief       For comparison: "Atomic" accesses using volatile
+ * @author      Marian Buschsieweke <marian.buschsieweke@ovgu.de>
+ *
+ * This header implements the `volatile_*()` family of functions of
```
Comment (kaspar030) [inline tests/sys_atomic_utils/volatile_utils.h:17]: While these might not guarantee atomic access, they do guarantee volatile access (as in, don't optimize, don't re-use a previously read value, ...). Maybe that's useful on its own?

File: sys/include/atomic_utils.h [hunk 3]
```diff
@@ -0,0 +1,1053 @@
...
+ * # Usage
+ *
+ * The atomic utilitys allow atomic access to regular variables.
+ *
+ *  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.c}
+ *  uint32_t global_counter;
+ *
+ *  void irq_handler(void)
+ *  {
+ *      // No need to use atomic access in IRQ handlers: RIOT does not
```
Comment (kaspar030) [inline sys/include/atomic_utils.h:24]: Isn't this a classic example of where compilers can optimize in unexpected ways, as soon as more code is added?

E.g.,

```
uint32_t global_counter;
 void irq_handler(void)
 {
   while (condition) {
     // do_register_magic
   }
   global_counter++;
}
```

The compiler is allowed to write to global_counter whenever it thinks is best.

File: sys/include/atomic_utils.h [hunk 4]
```diff
@@ -0,0 +1,1053 @@
...
+ *      // enable nested interrupts, so IRQ handlers will not be interrupted
```
Comment (kaspar030) [inline sys/include/atomic_utils.h:25]: (this is not entirely true anymore, eg., PendSV is now lower priority than the other ISRs)

File: sys/include/atomic_utils.h (hunk 3 above)
Comment (kaspar030) [reply sys/include/atomic_utils.h:24]: What I mean is, using atomic would ensure the compiler doesn't optimize around that write, correct?

(I know I argued before that *our* compilers on *our* architectures don't do those reorders... :innocent: )

File: cpu/arm7_common/include/atomic_utils_arch.h (hunk 1 above)
Comment (maribu) [reply cpu/arm7_common/include/atomic_utils_arch.h:1]: I think it might be useful to have the freedom to implement them completely independent from everyone else. Right now, I'm considering to implement them in inline assembly as LLVM has no convenience builtin functions that work on types not marked as `_Atomic`.

File: tests/sys_atomic_utils/volatile_utils.h (hunk 2 above)
Comment (maribu) [reply tests/sys_atomic_utils/volatile_utils.h:17]: Hmm. There might be use cases for that, but this is also provided by the `atomic_` functions (although potentially two instructions more expensive). It is often the case that one needs both the volatile access and the sequential order, but all current compilers move non-volatile code across volatile accesses.

I rather would not encourage programmer using this to safe 2 instructions, but risk bugs popping up due to incorrect use of them. But if you really want to export them, I could so and and I would add appropriate warnings. Hopefully people read those and think carefully before using them.

File: sys/include/atomic_utils.h (hunk 3 above)
Comment (maribu) [reply sys/include/atomic_utils.h:24]: We have (hopefully :-D) barriers in place at the beginning and end of each ISR. That way the compiler can reorder the writes within the ISR, but they should happen somewhere within the ISR. Also other optimizations like elimination of dead writes etc. can work properly, but the very last write cannot be assumed to be dead due to those barriers.

File: sys/include/atomic_utils.h (hunk 4 above)
Comment (maribu) [reply sys/include/atomic_utils.h:25]: Hmm, as a result every memory that is also accessed by PendSV must then use atomic accesses as well.

Let me rephrase this to be extra careful. If other ISRs do not touch the same variable, no issue would occur with nesting IRQs. And likely this is the default case.

File: cpu/arm7_common/include/atomic_utils_arch.h (hunk 1 above)
Comment (kaspar030) [reply cpu/arm7_common/include/atomic_utils_arch.h:1]: ok. I just thought that those impls mapping to "__atomic_store_N()" could be unified, but explicity also has its upsides.

File: tests/sys_atomic_utils/volatile_utils.h (hunk 2 above)
Comment (kaspar030) [reply tests/sys_atomic_utils/volatile_utils.h:17]: > I rather would not encourage programmer using this to safe 2 instructions, but risk bugs popping up due to incorrect use of them.

ok!

File: sys/include/atomic_utils.h (hunk 3 above)
Comment (kaspar030) [reply sys/include/atomic_utils.h]: right, I got confused with doing this in threads. nevermind

**Comment by kaspar030** [comment]
I'm looking at the benchmark results, ran on nrf52840dk. Overall this seems like a nice speedup over the c11 atomics.

Some things I don't understand:

```
| atom |     load |    8 |          7815 µs |         15627 µs |         10939 µs |
| atom |     load |   16 |          7815 µs |         10940 µs |         15627 µs |
| atom |     load |   32 |          7814 µs |         15626 µs |         10939 µs |
| atom |     load |   64 |          9377 µs |         10939 µs |         28127 µs |
```

Somehow the numbers look weird. I'd expect 64b to take longer than 32bit, for example.

```
| atom |      add |    8 |         12502 µs |         10940 µs |         25002 µs |                                                                
| atom |      add |   16 |         18752 µs |         12502 µs |         21877 µs |                                                                
| atom |      add |   32 |         10939 µs |         12502 µs |         26565 µs |                                                                
| atom |      add |   64 |         17190 µs |         10939 µs |         57814 µs |                                                                
...
| semi |      add |    8 |         12502 µs |         14065 µs |         20315 µs |
| semi |      add |   16 |         12501 µs |         18752 µs |         25002 µs |
| semi |      add |   32 |         10939 µs |         14064 µs |         20314 µs |
| semi |      add |   64 |         17189 µs |         10939 µs |         57814 µs |
```

Isn't semi_atomic supposed to be less atomic or fall back to full atomic, and should thus be faster?

**Comment by maribu** [comment]
> I'd expect 64b to take longer than 32bit, for example.

Yes, it should. I will double check if I messed something up in the benchmark. Maybe it is also something like cache efficiency depending on the position in ROM issue - e.g. like the issue with the STM32F767ZI on which faster code ran slower, except when the ISR vectors where of different size.

> Isn't semi_atomic supposed to be less atomic or fall back to full atomic, and should thus be faster?

If lock-free writes are supported (up to word size usually) and the same corresponding fully atomic would contain locks, it will be implemented faster and lock free (but less atomic). Otherwise, it will just be an alias for the corresponding atomic operation (and should be just as fast). The semi-atomic operations taking longer indicates an issue with the benchmark, or other effects like cache-efficiency due to ROM position outweigh the actual performance differences.

I have the same hardware at home, I should be able to reproduce the number tonight. Hopefully, I just messed up the benchmark. It would be a pitty, if the numbers are unreliable.

(This is by the way a good use case for AVRs :-) If the numbers don't make sense there, I surely fucked up the benchmark. Being a stupidly simple and, hence, highly predictable platform has its upsides :-))

**Comment by maribu** [comment]
> (This is by the way a good use case for AVRs :-) If the numbers don't make sense there, I surely fucked up the benchmark. Being a stupidly simple and, hence, highly predictable platform has its upsides :-))

On AVR, only the lock-free functions are slightly faster (`atomic_store_u8()` and `atomic_load_u8()`), literally all other functions (`atomic_store_u16()` and up, `atomic_load_u16()` and up, and all other `atomic_fetch_<op>_u<width>()`) implemented by `atomic_utils.h` took pretty much the same time.

The lock-free functions (which will also be lock free with real C11 atomics) take (+- some error) the same time for both atomic-utils and C11 atomics. So at least one thing is as expected.

I'm not so fluent with AVR assembly, but the ARM assembly makes sense. It seems like the machine code for functions in the atomic utils is pretty much what one would expect. I can't make any sense of the numbers now :-/

**Comment by maribu** [comment]
I amended the API for the bit clearing and setting, so that there is a dedicated type for a reference to a bit. This is useful to avoid computing the bit-band alias address of a given bit multiple times, as now the bit reference can be computed ahead of time and stored.

**Comment by kaspar030** [comment]
The benchmark seems borked.
I tried to isolate some numbers, so just chopped out everything but the semi add functions.
Previous full results (which were in between all the others):
```
| semi |      add |    8 |         14064 µs |         15627 µs |         26565 µs |
| semi |      add |   16 |         14064 µs |         20314 µs |         25001 µs |
| semi |      add |   32 |         10939 µs |         14064 µs |         26564 µs |
| semi |      add |   64 |         20314 µs |         15627 µs |         57814 µs |
```

In isolation:
```
Note: LOWER IS BETTER!

+------+----------+------+------------------+------------------+------------------+
| mode |       op | bits |         volatile |      atomic util |       c11 atomic |
+------+----------+------+------------------+------------------+------------------+
| semi |      add |    8 |         14064 µs |         18752 µs |         25002 µs |
| semi |      add |   16 |         12502 µs |         15627 µs |         20314 µs |
| semi |      add |   32 |         10939 µs |         14064 µs |         20314 µs |
| semi |      add |   64 |         20314 µs |         17190 µs |         59377 µs |
+------+----------+------+------------------+------------------+------------------+
```

Both are on an nrf52840dk.

**Comment by kaspar030** [comment]
I tried to isolate more:

https://gist.github.com/07353af10419c57e2fae7fac1bdeadc9

On the nrf52840dk, the first iteration takes a bit longer, then all are identical, for all types.
Looking at the assembly, it doesn't look like any atomic operation is called. Should it be optimized out in this case?

**Comment by maribu** [comment]
> Looking at the assembly, it doesn't look like any atomic operation is called.

For the atomic utils this is expected: Currently only the loads and stores up to the word size are provided by the platform specific implementation - and those should be regular loads and stores followed by (depending on platform) a data sync instruction. All other implementations rely on disabling IRQs during the read-modify-write sequence. (So there is still potential to optimize those operations for more capable MCUs, but that could be better done as follow up.)

The data sync might already be the explanation for why the time doesn't really increase as expected with bit size - I wouldn't be surprised if that is the major contributor to the time spent and the difference between the e.g. 32 bit and 64 bit is almost noise compared to that.

> Should it be optimized out in this case?

That would be a bug. The problem is that the compiler has absolutely no idea about concurrency, and thus has no chance to correctly detect when the atomically of atomic operations is not really needed. As a result, the compiler must not optimize the atomic accesses at all, otherwise it would also do so when this would actually be needed.

**Comment by maribu** [comment]
I found the issue :-)

**Comment by maribu** [comment]
Now the benchmark results do make sense

**Comment by kaspar030** [comment]
Re-ran on nrf52840dk:

```
main(): This is RIOT! (Version: 2020.10-devel-1529-g30a1e-atomic_utils)                                                   [31/3934]
Note: LOWER IS BETTER!                                                                                                             
                                                                                                                                   
+------+----------+------+------------------+------------------+------------------+
| mode |       op | bits |         volatile |      atomic util |       c11 atomic |
+------+----------+------+------------------+------------------+------------------+
| atom |    store |    8 |          6251 µs |         15627 µs |         10939 µs |
| atom |    store |   16 |          6251 µs |          9377 µs |         12502 µs |
| atom |    store |   32 |          6252 µs |         10939 µs |          9377 µs |
| atom |    store |   64 |          9376 µs |         15627 µs |         46877 µs |
| atom |     load |    8 |          7814 µs |         12502 µs |         14064 µs |
| atom |     load |   16 |          7814 µs |         10939 µs |         12501 µs |
| atom |     load |   32 |         12502 µs |         12501 µs |         10939 µs |
| atom |     load |   64 |          9376 µs |         15627 µs |         34377 µs |
| atom |      add |    8 |         14064 µs |         18752 µs |         20315 µs |
| atom |      add |   16 |         12502 µs |         20315 µs |         21877 µs |
| atom |      add |   32 |         15627 µs |         18752 µs |         25002 µs |
| atom |      add |   64 |         17189 µs |         23439 µs |         57814 µs |
| atom |      sub |    8 |         12502 µs |         25002 µs |         21877 µs |
| atom |      sub |   16 |         14064 µs |         18752 µs |         20314 µs |
| atom |      sub |   32 |         15627 µs |         21877 µs |         21877 µs |
| atom |      sub |   64 |         17189 µs |         23439 µs |         57814 µs |
| atom |       or |    8 |         10939 µs |         18752 µs |         20315 µs |
| atom |       or |   16 |         18752 µs |         20314 µs |         21877 µs |
| atom |       or |   32 |         15627 µs |         23439 µs |         20314 µs |
| atom |       or |   64 |         17189 µs |         25001 µs |         60939 µs |
| atom |      xor |    8 |         14064 µs |         18752 µs |         21877 µs |
| atom |      xor |   16 |         14064 µs |         25001 µs |         21877 µs |
| atom |      xor |   32 |         10939 µs |         18752 µs |         25002 µs |
| atom |      xor |   64 |         17190 µs |         25002 µs |         59376 µs |
| atom |      and |    8 |         14064 µs |         18752 µs |         21876 µs |                                        [1/3934]
| atom |      and |   16 |         14064 µs |         18752 µs |         21876 µs |                                                
| atom |      and |   32 |         10939 µs |         18752 µs |         20315 µs |
| atom |      and |   64 |         17189 µs |         23439 µs |         59376 µs |
| atom |      set |    8 |         10940 µs |         18752 µs |         25001 µs |
| atom |      set |   16 |         18752 µs |         25002 µs |         21877 µs |
| atom |      set |   32 |         10939 µs |         21877 µs |         26565 µs |
| atom |      set |   64 |         17189 µs |         25001 µs |         60939 µs |
| atom |    clear |    8 |         10939 µs |         18752 µs |         20314 µs |
| atom |    clear |   16 |         14064 µs |         21876 µs |         20314 µs |
| atom |    clear |   32 |         10939 µs |         18752 µs |         20315 µs |
| atom |    clear |   64 |         18752 µs |         31251 µs |         59377 µs |
| semi |      add |    8 |         18752 µs |         15627 µs |         20314 µs |
| semi |      add |   16 |         14064 µs |         15626 µs |         20314 µs |
| semi |      add |   32 |         10939 µs |         14064 µs |         20314 µs |
| semi |      add |   64 |         17189 µs |         23439 µs |         57814 µs |
| semi |      sub |    8 |         12502 µs |         18752 µs |         25001 µs |
| semi |      sub |   16 |         17189 µs |         14064 µs |         20314 µs |
| semi |      sub |   32 |         10939 µs |         14064 µs |         20314 µs |
| semi |      sub |   64 |         17189 µs |         23439 µs |         57814 µs |
| semi |       or |    8 |         12501 µs |         17189 µs |         25002 µs |
| semi |       or |   16 |         14064 µs |         14064 µs |         20314 µs |
| semi |       or |   32 |         10939 µs |         14064 µs |         28127 µs |
| semi |       or |   64 |         17190 µs |         25002 µs |         60939 µs |
| semi |      xor |    8 |         10940 µs |         14064 µs |         20314 µs |
| semi |      xor |   16 |         12502 µs |         17190 µs |         26564 µs |
| semi |      xor |   32 |         10939 µs |         14064 µs |         20314 µs |
| semi |      xor |   64 |         17189 µs |         25002 µs |         59377 µs |
| semi |      and |    8 |         10939 µs |         14064 µs |         20314 µs |
| semi |      and |   16 |         10939 µs |         14064 µs |         20314 µs |
| semi |      and |   32 |         10940 µs |         18752 µs |         25001 µs |
| semi |      and |   64 |         17189 µs |         23439 µs |         59376 µs |
+------+----------+------+------------------+------------------+------------------+
```

**Comment by kaspar030** [comment]
Somehow the benchmark is still not stable.
This is an isolated test:

```
main(): This is RIOT! (Version: 2020.10-devel-1529-g30a1e-atomic_utils)
Note: LOWER IS BETTER!

+------+----------+------+------------------+------------------+------------------+
| mode |       op | bits |         volatile |      atomic util |       c11 atomic |
+------+----------+------+------------------+------------------+------------------+
| semi |      add |    8 |         12501 µs |         14064 µs |         26564 µs |
| semi |      add |   16 |         14064 µs |         15626 µs |         25002 µs |
| semi |      add |   32 |         10939 µs |         14064 µs |         20314 µs |
| semi |      add |   64 |         17189 µs |         23439 µs |         59377 µs |
+------+----------+------+------------------+------------------+------------------+
```

This is the same but together with the others:

```
| semi |      add |    8 |         18752 µs |         15627 µs |         20314 µs |
| semi |      add |   16 |         14064 µs |         15626 µs |         20314 µs |
| semi |      add |   32 |         10939 µs |         14065 µs |         20315 µs |
| semi |      add |   64 |         17189 µs |         23440 µs |         57815 µs |
```

We can write those differences off as different optimization opportunities by the compiler...

**Comment by maribu** [comment]
On AVR the results are pretty stable. For e.g. atomic loads and stores I get some value between 50020 µs and 50084 µs for both atomic utils and C11 atomics all of the time. (So when they differ, it is due to noise - which makes sense, as the are implemented in the very same way.) So we should take all numbers below with an uncertainty of at least 64 µs.

Apart from that the only surprise is that volatile loads and stores perform slower than atomic loads and stores for word sized accesses. But everything else behaves as if it would be a text book result.

```
+------+----------+------+------------------+------------------+------------------+
| mode |       op | bits |         volatile |      atomic util |       c11 atomic |
+------+----------+------+------------------+------------------+------------------+
| atom |    store |    8 |         75096 µs |         50020 µs |         50084 µs |
| atom |    store |   16 |         87592 µs |         93836 µs |        206336 µs |
| atom |    store |   32 |         87592 µs |        118836 µs |        250152 µs |
| atom |    store |   64 |        137592 µs |        168900 µs |        368900 µs |
| atom |     load |    8 |         75092 µs |         50020 µs |         50020 µs |
| atom |     load |   16 |         87592 µs |         93836 µs |        187588 µs |
| atom |     load |   32 |         87532 µs |        118836 µs |        462780 µs |
| atom |     load |   64 |        137596 µs |        168904 µs |        275152 µs |
| atom |      add |    8 |        118780 µs |        100088 µs |        206404 µs |
| atom |      add |   16 |        100092 µs |        131336 µs |        256404 µs |
| atom |      add |   32 |        162592 µs |        193900 µs |        587780 µs |
| atom |      add |   64 |        381472 µs |        412716 µs |       1019288 µs |
| atom |      sub |    8 |        118840 µs |        100084 µs |        212584 µs |
| atom |      sub |   16 |        100096 µs |        131336 µs |        262648 µs |
| atom |      sub |   32 |        162592 µs |        193836 µs |        600348 µs |
| atom |      sub |   64 |        381408 µs |        412780 µs |       1019224 µs |
| atom |       or |    8 |        118844 µs |        100084 µs |        206336 µs |
| atom |       or |   16 |         93840 µs |        125084 µs |        256400 µs |
| atom |       or |   32 |        143840 µs |        175152 µs |        587784 µs |
| atom |       or |   64 |        243908 µs |        275148 µs |        687848 µs |
| atom |      xor |    8 |        118840 µs |        100020 µs |        206400 µs |
| atom |      xor |   16 |        100028 µs |        131336 µs |        256400 µs |
| atom |      xor |   32 |        150092 µs |        181400 µs |        587784 µs |
| atom |      xor |   64 |        243908 µs |        275148 µs |        687848 µs |
| atom |      and |    8 |        118840 µs |        100020 µs |        206400 µs |
| atom |      and |   16 |        100028 µs |        131336 µs |        256400 µs |
| atom |      and |   32 |        162596 µs |        193904 µs |        587780 µs |
| atom |      and |   64 |        243908 µs |        275152 µs |        687844 µs |
| atom |      set |    8 |        118844 µs |        100024 µs |        206400 µs |
| atom |      set |   16 |        143844 µs |        125088 µs |        243900 µs |
| atom |      set |   32 |        193908 µs |        175084 µs |        575280 µs |
| atom |      set |   64 |        243912 µs |        275152 µs |        675344 µs |
| atom |    clear |    8 |        118840 µs |        100088 µs |        200084 µs |
| atom |    clear |   16 |        143844 µs |        125088 µs |        243904 µs |
| atom |    clear |   32 |        193908 µs |        175084 µs |        581596 µs |
| atom |    clear |   64 |        243908 µs |        275152 µs |        675344 µs |
| semi |      add |    8 |        118844 µs |         68836 µs |        200084 µs |
| semi |      add |   16 |        100028 µs |        131336 µs |        256404 µs |
| semi |      add |   32 |        162592 µs |        193900 µs |        587780 µs |
| semi |      add |   64 |        381472 µs |        412716 µs |       1019292 µs |
| semi |      sub |    8 |        118844 µs |         68836 µs |        206336 µs |
| semi |      sub |   16 |        100092 µs |        131336 µs |        262652 µs |
| semi |      sub |   32 |        162592 µs |        193836 µs |        600348 µs |
| semi |      sub |   64 |        381412 µs |        412784 µs |       1019228 µs |
| semi |       or |    8 |        118844 µs |         68836 µs |        200084 µs |
| semi |       or |   16 |         93776 µs |        125084 µs |        256400 µs |
| semi |       or |   32 |        143844 µs |        175084 µs |        587844 µs |
| semi |       or |   64 |        243844 µs |        275216 µs |        687844 µs |
| semi |      xor |    8 |        118844 µs |         68836 µs |        200088 µs |
| semi |      xor |   16 |        100092 µs |        131336 µs |        256404 µs |
| semi |      xor |   32 |        150092 µs |        181336 µs |        587848 µs |
| semi |      xor |   64 |        243908 µs |        275152 µs |        687848 µs |
| semi |      and |    8 |        118844 µs |         68836 µs |        212584 µs |
| semi |      and |   16 |        100096 µs |        131336 µs |        256400 µs |
| semi |      and |   32 |        162596 µs |        193904 µs |        587780 µs |
| semi |      and |   64 |        243908 µs |        275152 µs |        687848 µs |
+------+----------+------+------------------+------------------+------------------+
```

Maybe we should add a gnuplot command stream output mode to our benchmarks? :laughing: I really would like to see those graphically :-)

**Comment by kaspar030** [comment]
maybe squash this?

**Comment by maribu** [comment]
Squashed and renamed `tests/bench_atomic_util` to `tests/bench_sys_atomic_utils`.

**Review by kaspar030 — APPROVED** [review]
ACK.

Assembly looks OK now, the whole concept is nice. Let's go!

**Comment by maribu** [comment]
Diff since ACK: https://github.com/RIOT-OS/RIOT/compare/bfd25911991ca297a1f10fc0ba52286cf561a0ee..fb838b74bf09d453cf0e5ad0a12984f46d17c402

**Comment by kaspar030** [comment]
re-ACK. I gave the PR a last complete look. Ran benchmarks, test and unittests on nrf52840dk. No issues.

Let's give murdock another run with tests in samr21-xpro and esp32.

File: tests/unittests/tests-atomic_utils/tests-atomic_utils.c [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+#include <string.h>
+
+#include "embUnit.h"
+#include "tests-atomic_utils.h"
+
+#include "atomic_utils.h"
+#include "random.h"
+
+#define ENABLE_DEBUG (0)
```
Comment (fjmolinas) [inline tests/unittests/tests-atomic_utils/tests-atomic_utils.c:28]:
```suggestion
#define ENABLE_DEBUG 0
```

**Comment by fjmolinas** [comment]
All tests seem to be ok except `tests/pkg_libhydrogen/samr21-xpro:llvm, seems like its OK to squash @maribu

File: tests/unittests/tests-atomic_utils/tests-atomic_utils.c (hunk 5 above)
Comment (maribu) [reply tests/unittests/tests-atomic_utils/tests-atomic_utils.c:28]: Good catch! :-)

**Comment by chrysn** [comment]
In the comparison of IRQs vs. atomics for performance, was this evaluated only with SEQ_CST or also with ACQUIRE and RELEASE? After all, some say (citation needed, I've read an article but can't find it any more) that if you do SEC_CST you're almost always asking more of the platform than you need.

**Comment by maribu** [comment]
> After all, some say (citation needed, I've read an article but can't find it any more) that if you do SEC_CST you're almost always asking more of the platform than you need.

tl;dr: No, I currently don't intent to extent this API to offer other guarantees than sequentially consistence. IMO that would gain only limited benefit in our context, but we would pay for that with increased complexity and confused developers.

---

The comparison is only for sequentially consistent. IMO, adding more variants lead to more complexity and confusion. I already had quite a few highly emotional, exhausting, and lengthy discussions on whether just adding `volatile` is enough for inter-context-communication (thread-to-thread or thread-to-isr) via shared memory. I think getting the "safest" atomic access variant as fast as possible is time better spent on than on reasoning on whether relaxing the guarantees is possible, documenting it, and discussing it.

Note that overhead for sequentially consistent atomic accesses in the context of embedded systems is often *significantly* less than it is in multi-core superscalar out-of-order CPUs. First, disabling IRQs briefly is ***much*** cheaper than a context switch to the OS and back for operations that lack lock-free implementations. And e.g. a lock on the memory bus that might cause other CPU cores to block during loads and stores from/to RAM is something we also don't need to worry about. (When we add support for multi-core MCUs, I think a high level primitive like an yet to implement `mutli_core_mutex_t` and carefully trying to avoid inter-core communication wherever possible is the better route to go.)

**Comment by aabadie** [comment]
Hmm, I broke Murdock when merging this one: the unittests don't fit anymore in arm7 boards (msba2 and mcb2388). I suggest to move the atomic function unittest to its own test application (instead of simply excluding these boards from the unittests) ?

**Comment by aabadie** [comment]
Let's see if #15438 fixes the problem
